services/powerbi/powerbi.go

```diff
@@ -0,0 +1,207 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/Azure/adx-automation-services/services/powerbi/kubeutils"
+	"github.com/Azure/adx-automation-services/services/powerbi/logging"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+)
+
+type PowerBIService interface {
+	Refresh(context.Context, string, int64) error
+}
+
+type powerBIService struct{}
+
+const (
+	emptyProductError = "'product' is not set"
+	skipTemplate      = "secret does not have a `%s` key. Skipping Power BI refresh"
+	groupKey          = "powerbi.group"
+	datasetKey        = "powerbi.dataset"
+	maxAttempts       = 20
+	sleepTime         = time.Second * 30
+)
+
+// Refresh refreshes a PowerBI dataset.
+// See https://powerbi.microsoft.com/en-us/blog/announcing-data-refresh-apis-in-the-power-bi-service/
+func (powerBIService) Refresh(_ context.Context, product string, runID int64) error {
+	logging.LogInfo("starting with PowerBI refresh")
+	// get parameters
+	if product == "" {
+		return errors.New(emptyProductError)
+	}
+	secret, err := kubeutils.TryCreateKubeClientset().
+		CoreV1().
+		Secrets(kubeutils.GetCurrentNamespace("a01-prod")).
+		Get(product, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get the kubernetes secret: %v", err)
+	}
+
+	group, ok := secret.Data[groupKey]
+	if !ok {
+		return fmt.Errorf(skipTemplate, groupKey)
+	}
+	dataset, ok := secret.Data[datasetKey]
+	if !ok {
+		return fmt.Errorf(skipTemplate, datasetKey)
+	}
+	logging.LogInfo(fmt.Sprintf("runID: %d | product: %s | group: %s | dataset: %s",
+		runID,
+		product,
+		string(group),
+		string(dataset)))
+
+	// get client and authorizer
+	client := autorest.NewClientWithUserAgent("")
+	authorizer, err := getAuth()
+	if err != nil {
+		return fmt.Errorf("failed to get authorizer: %v", err)
+	}
+	client.Authorizer = authorizer
+
+	uri := fmt.Sprintf("https://api.powerbi.com/v1.0/myorg/groups/%s/datasets/%s/refreshes", string(group), string(dataset))
```

Comment:
You may want to make this configurable.

Reply:
All datasets follow this pattern.
Pattern and group values should be part of the product secret

Reply:
ok.